packages/bigquery/src/index.js

```diff
@@ -57,16 +57,30 @@ var Table = require('./table.js');
  *
  * @resource [What is BigQuery?]{@link https://cloud.google.com/bigquery/what-is-bigquery}
  *
+ * The apiEndpoint from options will set the host. If not set, the
+ * `GOOGLE_CLOUD_BIGQUERY_ENDPOINT` environment variable is honored,
+ * otherwise the actual API endpoint will be used.
+ *
  * @param {object} options - [Configuration object](#/docs).
+ * @param {object} options.apiEndpoint - Override the default API endpoint
+ * to reach BigQuery.
  */
 function BigQuery(options) {
   if (!(this instanceof BigQuery)) {
     options = common.util.normalizeArguments(this, options);
     return new BigQuery(options);
   }
 
+  var basePath = '/bigquery/v2';
+
+  var baseInfo = common.util.determineBaseUrl(
+      options,
+      'GOOGLE_CLOUD_BIGQUERY_ENDPOINT',
+      'https://www.googleapis.com');
+
   var config = {
-    baseUrl: 'https://www.googleapis.com/bigquery/v2',
+    baseUrl: baseInfo.apiEndpoint + basePath,
+    customEndpoint: baseInfo.customEndpoint,
```

Comment:
What do you think about having `common.util.determineBaseUrl` called from the `common.Service` constructor? We could pass in config data to Service, like `config.defaultEndpoint` and `config.customEndpointEnvVar` (or something similar), and let the Service constructor handle the rest.

Does that seem possible, or am I overlooking any parts of the codebase where that would be a problem?

Reply:
Actually, looking at the [current](https://github.com/GoogleCloudPlatform/google-cloud-node/blob/master/packages/bigtable/src/index.js#L318) implementation of bigtable, if there is a custom endpoint it overrides every baseUrl. We can bring this down to service.js like this:

```
  ...

  var baseInfo = util.determineBaseUrl(
    options,
    config.environmentVariables,
    config.defaultApiEndpoint,
    true
  );

  ...

  Object.keys(protoServices).forEach(function(name) {
    var protoConfig = protoServices[name];
    var service = self.loadProtoFile_(protoConfig, config);

    self.protos[name] = service;

    if (baseInfo.customEndpoint) {
      service.baseUrl = baseInfo.apiEndpoint;
    } else if (!protoConfig.baseUrl) {
      service.baseUrl = baseInfo.apiEndpoint;
    }
  });

...

```

Essentially overriding only if there is a custom endpoint or if for whatever reason, the protoService doesn't already have a baseUrl defined

Reply:
Ah, nice catch.

> Essentially overriding only if there is a custom endpoint or if for whatever reason, the protoService doesn't already have a baseUrl defined

That sounds fine, let's go with that.

Reply:
Perfect. I'll work this changes this morning.